src/kaleidoscope_internal/typed_plugins.h

```diff
@@ -0,0 +1,226 @@
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2013-2018  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <stdint.h>
+
+#include "kaleidoscope/macro_map.h"
+
+// The following is template code that generates an array of pointers of
+// a specific type of plugins (e.g. LEDMode's) at compiletime and can be
+// stored in PROGMEM.
+
+namespace kaleidoscope {
+namespace internal {
+namespace typed_plugins {
+
+// The TypedPluginArray class stores pointers to plugins of a specific
+// type in a recursive fashion.
+//
+template<typename TargetType__,
+         bool IsAppropriateType__,
+         bool...MoreTypeInfo__>
+class TypedPluginArray {
+
+ public:
+
+  typedef TypedPluginArray<TargetType__, MoreTypeInfo__...> NestedArray;
+
+  template<typename...MorePlugins__>
+  constexpr TypedPluginArray(TargetType__ *entry, MorePlugins__...more_plugins)
+    : entry_(entry),
+      nested_array_(more_plugins...) {}
+
+  static constexpr uint8_t n_entries
+    = NestedArray::n_entries + 1;
+
+  typedef TargetType__ ContentType;
+
+ private:
+  TargetType__ *entry_;
+  NestedArray nested_array_;
+} __attribute__((packed)); // Make sure that there are no padding
+// bytes added by the compiler.
+// This is important to let the class
+// have the same layout as a POD array.
+
+
+template<typename TargetType__,
+         bool...MoreTypeInfo__>
+class TypedPluginArray<TargetType__,
```

Comment:
When it encounters `TypedPluginArray<Foo, true, true>`, how does the compiler decide which template class to use? It seems like both of them match.

Reply:
The compiler will always prefer the specialization, if available. That enables conditional clauses based on types.

Reply:
In that case, does the second one only get used if there are two template parameters? If so, why did you use a parameter pack instead of just a single `bool`?

Reply:
The following implementations of the `TypedPluginArray` class will be instantiated.

`TypedPluginArray<Foo, true, ... more args ...>`: line 34
`TypedPluginArray<Foo, false, ... more args ...>`: line 62
`TypedPluginArray<Foo, true>`: line 87
`TypedPluginArray<Foo, false>`: line 104

The template classes are instantiated recursively. That's why we need two additional specializations (lines 87 and 104) to end recursion. If a `bool` was used instead of a parameter pack, `TypedPluginArray` could not support more than two plugins.

Reply:
So, the template on line 62 never actually gets used, because, depending on the value of its second parameter, it is overridden by one of the templates on lines 87 and 104?

Reply:
Oh – no, I was wrong. The one on line 62 overrides the one on line 34 if the value of the second parameter is `false`?

Reply:
The problem I was having understanding this was that template specialization syntax is not what I would expect it to be. In a non-specialized template, the parameters that are used for instantiation appear after the `template` keyword, and before the class name, but for specialization, they appear after the class name, and there are fewer parameters listed after `template` (and the order appears to change if anything but the last parameter is the specialized one). I think I've got that part now.